module/Application/src/Application/Service/ProjectRetrievalService.php

```diff
@@ -23,44 +24,68 @@ class ProjectRetrievalService
         $this->http = new Client(null, $options);
-        $this->validator = $validator;
+        $this->schemaValidator = $validator;
     }
 
     /**
      * @param $url
      * @return \Application\Project
      * @throws \Zend\Http\Client\Exception\RuntimeException
-     * @throws ProjectRetrievalException
-     * @throws ProjectValidationException
+     * @throws RetrievalException
+     * @throws ValidationException
      */
     public function getByUrl($url)
     {
         $response = $this->http->setUri($url)->send();
         if (!$response->isSuccess()) {
-            throw new ProjectRetrievalException('Non-successful response code', $response->getStatusCode());
+            throw new RetrievalException('Non-successful response code', $response->getStatusCode());
         }
 
-        return $this->parseResponse($response->getBody());
+        return $this->buildFromResponse($response->getBody(), $url);
     }
 
     /**
      * @param string $json_string
+     * @param string $url needed for validation
      * @return Project
-     * @throws ProjectValidationException
+     * @throws RetrievalException
+     * @throws ValidationException
      */
-    protected function parseResponse($json_string) {
-        $rawObject = json_decode($json_string, JSON_OBJECT_AS_ARRAY);
-        if (!$rawObject) {
-            throw new ProjectRetrievalException('Could not parse JSON', static::JSON_PARSE_FAILURE);
+    protected function buildFromResponse($json_string, $url)
+    {
+        $rawArray = json_decode($json_string, JSON_OBJECT_AS_ARRAY);
+        if (!$rawArray) {
+            throw new RetrievalException('Could not parse JSON', static::JSON_PARSE_FAILURE);
+        }
+
+        if (!$this->schemaValidator->isValid($rawArray)) {
+            throw new ValidationException($this->schemaValidator->getMessages(), static::SCHEMA_VALIDATION_FAILURE);
         }
 
-        if (!$this->validator->isValid($rawObject)) {
-            throw new ProjectValidationException($this->validator->getMessages(), static::SCHEMA_VALIDATION_FAILURE);
+        if (!$this->isUrlParent($rawArray['repository'], $url)) {
+            throw new ValidationException(
+                ['repository' => 'Retrieved file isn\'t in this repo'],
+                static::PARENT_VALIDATION_FAILURE
+            );
         }
 
         $project = new Project();
-        $project->exchangeArray($rawObject);
+        $project->exchangeArray($rawArray);
 
         return $project;
     }
+
+    /**
+     * Validates that a child URL is in a subdirectory (e.g. within the base repository) of
+     * the parent URL. Immediately rejects parent URLs with ..'s in them due to the potential
+     * for repo hijacking.
+     *
+     * @param string $parent_url
+     * @param string $child_url
+     * @return bool
+     */
+    protected function isUrlParent($parent_url, $child_url)
+    {
+        return strpos($parent_url, '..') === false && strpos($child_url, $parent_url) === 0;
```

Comment:
@dshafik is already validating this piece. No need to double-dip. Removed this bit.